tket/src/serialize/pytket/circuit.rs

```diff
@@ -0,0 +1,430 @@
+use std::sync::Arc;
+
+use hugr::core::HugrNode;
+use hugr::hugr::hugrmut::HugrMut;
+use hugr::ops::handle::NodeHandle;
+use hugr::ops::{OpParent, OpTag, OpTrait};
+use hugr::{Hugr, HugrView, Node};
+use hugr_core::hugr::internal::HugrMutInternals;
+use itertools::Itertools;
+use rayon::iter::{IntoParallelRefIterator, IntoParallelRefMutIterator, ParallelIterator};
+use tket_json_rs::circuit_json::{Command as PytketCommand, SerialCircuit};
+
+use crate::serialize::pytket::decoder::PytketDecoderContext;
+use crate::serialize::pytket::opaque::SubgraphId;
+use crate::serialize::pytket::{
+    default_encoder_config, DecodeInsertionTarget, DecodeOptions, EncodeOptions, PytketDecodeError,
+    PytketDecodeErrorInner, PytketDecoderConfig, PytketEncodeError, PytketEncoderContext,
+};
+use crate::Circuit;
+
+use super::opaque::OpaqueSubgraphs;
+
+/// An encoded pytket circuit that may be linked to an existing HUGR.
+///
+/// Tracks correspondences between references to the HUGR in the encoded
+/// circuit, so we can reconstruct the HUGR if needed.
+///
+/// Serial circuits in this structure are intended to be transient, only alive
+/// while this structure is in memory. To obtain a fully standalone pytket
+/// circuit that can be used independently, and stored permanently, use
+/// [`EncodedCircuit::new_standalone`] or call
+/// [`EncodedCircuit::ensure_standalone`].
+#[derive(Debug, Clone)]
+pub struct EncodedCircuit<Node: HugrNode> {
+    /// Circuits encoded from independent dataflow regions in the HUGR.
+    ///
+    /// These correspond to sections of the HUGR that can be optimized
+    /// independently.
+    circuits: HashMap<Node, EncodedCircuitInfo>,
+    /// Sets of subgraphs in the HUGR that have been encoded as opaque barriers
+    /// in the pytket circuit.
+    ///
+    /// Subcircuits are identified in the barrier metadata by their ID in this
+    /// vector. See [`SubgraphId`].
+    opaque_subgraphs: OpaqueSubgraphs<Node>,
+}
+
+/// Information stored about a pytket circuit encoded from a HUGR region.
+#[derive(Debug, Clone)]
+pub(super) struct EncodedCircuitInfo {
+    /// The serial circuit encoded from the region.
+    pub serial_circuit: SerialCircuit,
+    /// A subgraph of the region that does not contain any operation encodable
+    /// as a pytket command, and hence was not encoded in [`serial_circuit`].
+    #[expect(unused)]
+    pub extra_subgraph: Option<SubgraphId>,
+    /// List of parameters in the pytket circuit in the order they appear in the
+    /// hugr input.
+    ///
+    /// We require this to correctly reconstruct the input order in the reassembled hugr,
+    /// since parameters in pytket are unordered.
+    pub input_params: Vec<String>,
+    /// List of output parameter expressions found at the end of the encoded region.
+    //
+    // TODO: The decoder does not currently connect these.
+    pub output_params: Vec<String>,
+}
+
+impl EncodedCircuit<Node> {
+    /// Encode a HugrView into a [`EncodedCircuit`].
```

Comment:
```suggestion
    /// Encode a Hugr into a [`EncodedCircuit`].
```
I mean, Node=Node and AsRef and AsMut is close enough to Hugr, right. (But I'm a little surprised by the need for AsMut - presumably some trait needs this? I mean, you can't actually mutate through the `&Circuit<H>`, can you?)

Reply:
Yeah, AsMut<Hugr> is pretty much a Hugr.

The limitation are the Hugr builder traits, that require AsMut/AsRef. That's why decoding is only done on Hugrs.